### DISCUSSION TITLE
FYI: Prograph CPX

### DISCUSSION THREAD

**Discussion by everythingability** [FlydeHQ/flyde#104] [General]
Hello...

I'm just reading the Flyde documentation and many aspects of it remind me of a visual programming tool from the 90s called Prograph CPX.  I wondered if you were aware of it at all... and MOST IMPORTANTLY if you could replicate/rip off the ideas of ...

Scrunching --- where methods or bit of code get grouped and become one node that you need to zoom into...That's the first part but then with these sort of boxes of code (with inputs and outputs) you could then run commands like ...

- ** Local-to-opers **- which made that box of nodes into a fully fledged function in it's own right, rather than just a tidying up of some** code...
- **Opers-to-local** - did the opposite, sort of dropping a functions code into your code 

I only mention this because these simple abilities REALLY helped to avoid spagetti code and loads of visual tools end up with spagetti code. 

Here is a video of me waxing lyrical nostaligicly explaining some of these features (really badly) , but still, I thought you may be interested.

https://youtu.be/MtECJw59elc?si=B7Yc6MfrdOdMwItr&t=1430

Looking forward to getting deeper into Flyde...

**Comment by GabiGrin** [comment]
Never heard of it TBH! Will look  into and watch your videos asap.

Seems super interesting. The zoom in feature sounds like https://xai-primer.com/tool/

Thanks for sharing and looking forward to further thoughts and feedback 🙏

**Comment by shabir-mp** [comment]
Very good project 👍. If I may request, I want the program to be able to visualize all programming languages. 😁

**Comment by everythingability** [comment]
Ah no, @GabiGrin, Prograph CPX didn't have a zoom feature but I am a massive fan of proper ZUIs, which had a purple patch in the nineties and noughties, mainly as prototypes, then fizzled out. (Thanks for the link).

@shabir-mp If you mean you want to try out Prograph, you will need to emulate MacOS System 9 (or 8) with Sheepshaver. But there was a free copy of Prograph that still runs on Mac called Marten. https://www.andescotia.com/products/marten/ So you can download that and try "editing code as you write it" ... insanely brilliant.

Unfortunately, the company that made Prograph suffered when Apple were in dire straits, they attempted to jump ship to Windows and the final nail was when the founder and genius behind it died (I believe).

**Comment by GabiGrin** [comment]
@everythingability I finally got around to watching your video and it completely blew my mind! It's the closest I've seen to Flyde ever. 
THANK YOU! The fact it also had an object-oriented visual editor and had C-matching performance is jaw-dropping.

Happy to say Flyde supports grouping and ungrouping nodes together, and will soon also support extracting a grouped node into its own file - check out here - https://www.youtube.com/watch?v=Eq7fxOafznQ

Looking forward to your thoughts, and thanks again for the reference. It did push the option for a "top-bottom design" to the top of my backlog.

**Comment by everythingability** [comment]
I'm glad you liked it, and "got it".  Not many people do, but I hoped you
might :-)

Another feature (accidental) of Prograph, was that you could (and I used
to) do pair programming... in a way you can't do easily with text.  *Visual*
is different in ways that are hard to explain.
